x-pack/legacy/plugins/lens/public/debounced_component/debounced_component.test.tsx

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import React from 'react';
+import { mountWithIntl as mount } from 'test_utils/enzyme_helpers';
+import { debouncedComponent } from './debounced_component';
+
+describe('debouncedComponent', () => {
+  test('immediately renders', () => {
+    const TestComponent = debouncedComponent(({ title }: { title: string }) => {
+      return <h1>{title}</h1>;
+    });
+    expect(mount(<TestComponent title="hoi" />).html()).toMatchInlineSnapshot(`"<h1>hoi</h1>"`);
+  });
+
+  test('debounces changes', async () => {
+    const TestComponent = debouncedComponent(({ title }: { title: string }) => {
+      return <h1>{title}</h1>;
+    }, 1);
+    const component = mount(<TestComponent title="there" />);
+    component.setProps({ title: 'yall' });
+    expect(component.html()).toMatchInlineSnapshot(`"<h1>there</h1>"`);
+    await new Promise(r => setTimeout(r, 1));
+    expect(component.html()).toMatchInlineSnapshot(`"<h1>yall</h1>"`);
+  });
+});
```

Comment:
I find the snapshot to be more effective than the examples you gave. Maybe we can discuss.

Reply:
The reason I say they're weak is that snapshots aren't testing the behavior of the system. If I change the behavior here, I want tests to fail. Instead of tests failing with a specific message, it says that the snapshots have changed and I see a diff that I have to interpret myself. There's no assertion of specific behavior.

For example, the issue with the SeriesType in the XY chart was made worse by snapshot testing. We were snapshotting the rendered chart, but without any assertions that the right series was rendered. So until I fixed it, we were rendering the `horizontal_bar` chart as a vertical area chart.

My preferred solution is combination, like I did for the SeriesType issue:

```typescript
    test('it renders stacked horizontal bar', () => {
      const { data, args } = sampleArgs();

      const component = shallow(
        <XYChart data={data} args={{ ...args, seriesType: 'horizontal_bar_stacked' }} />
      );
      expect(component).toMatchSnapshot();
      expect(component.find(BarSeries)).toHaveLength(1);
      expect(component.find(BarSeries).prop('stackAccessors')).toHaveLength(1);
      expect(component.find(Settings).prop('rotation')).toEqual(90);
});
```

That kind of test will catch both unexpected additions and fail fast if the required behavior changes.